### PR TITLE
End-to-end integration tests for WireGuard, OpenVPN, OpenConnect, Tor, Stunnel

### PR DESCRIPTION
**Pull request by cpu**
## Notes

_Note: at this point `vagrant up` and running the `streisand-client` VM against the `streisand-host` VM fails. I'm punting on this for now to focus on testing a real cloud instance as a starting point_.

_Note 2: The OpenVPN test doesn't yet exercise the `direct-udp` or `combined` OpenVPN client profiles._

~_Note 3: This PR depends on #1069 - You need to provision your server with a hybrid of this branch and #1069 until that has been merged to master in order for the needed `.mobileconfig` files to be present when running the OpenConnect test_~

~_Note 4: @alimakki points out that the OpenConnect tests are failing right now. The xpath for extracting the PCKS12 password from the XML mobileconfig broke with some changes that were made to the mobileconfig structure. I'll address this and remove the `status/revisions-needed` label once done_~

## Usage & Testing
To test this PR you'll want to provision a Streisand server after editing `global_vars/vars.yml` to set `streisand_client_test: yes`. This is required to write some additional files on the server that the client test can access (e.g. to know the randomly generated vpn client names, tor hidden service name).  After the server completes provisioning prepare to run the e2e integration tests from a local VM by setting `STREISAND_SERVER_IP` to your test server's IP and `STREISAND_PASSWORD` to the generated gateway password for the remote server. Then run:

     STREISAND_SERVER_IP='XX.XX.XX.XX'; STREISAND_PASSWORD='SomeRandomPassword;  printf "$STREISAND_SERVER_IP\n$STREISAND_PASSWORD\n" | ./tests/remote_test.sh

Feedback welcome! I'm going to continue working on some other client tests.

## WireGuard
The new improved WireGuard integration test now properly tests that a successful handshake between the client and server occurs. It will also check that the WireGuard DNS is added and removed from `/etc/resolv.conf` successfully. The integration test checks each of the generated WireGuard profiles to ensure each works correctly.

## OpenVPN
The new OpenVPN integration test checks all of the randomly generated client profiles across the different OpenVPN configurations we offer. Similar to the Wireguard tests it will check that a successful connection can be made, that the DNS is updated as expected, and that a remote website is accessible with the VPN up. At present only the `direct` and `sslh` type profiles are tested. The `direct-udp` and `combined` profiles aren't tested - they weren't working for me and I haven't yet determined if it's due to my peculiar home network or if there are bugs at work.

## OpenConnect
The new OpenConnect integration test checks all of the randomly generated PKCS12 client profiles. It learns the PKCS12 passwords from the AnyConnect `.mobileconfig` XML files. Similar to the other services we test that DNS is modified as expected & that an Internet resource is accessible over the service.

## Tor
The Tor integration test checks that the obfs4proxy relay on a Streisand service works correctly as a bridge relay for the test client. We also test that the Streisand hidden service is accessible, enforcing basic auth, and serving the homepage to users that know the username/password.

## Stunnel
The Stunnel integration test checks that the stunnel4 client configuration works as advertised & that the OpenVPN profiles that target the stunnel4 proxy work too. This test imports the OpenVPN profile testing mechanics to test that the DNS is updated & the overall Stunnel wrapped VPN connection works.

### CODE REVIEW

**Comment by cpu** [comment]
I broke this with the multi WG support that landed in master. Closing temporarily while I work that out (& subsequently beg/pester for reviews)

**Comment by cpu** [comment]
This PR depends on https://github.com/StreisandEffect/streisand/pull/1069 - It shouldn't be merged before that lands in master!

**Comment by cpu** [comment]
@jlund @alimakki @nopdotcom @nickolasclarke Between the lot of you can someone do a review pass of this PR in the next little while? It minimally affects the main (non-test) codebase. If you could even ignore the test commits and just sign off on the parts modifying the main code I'd be happy with merging as-is. Nothing will be using the end to end tests yet. Thanks!

File: playbooks/roles/openconnect/templates/instructions.md.j2 [hunk 1]
```diff
@@ -13,7 +13,7 @@ OpenConnect is an extremely high-performance and lightweight VPN server that als
   * [iOS](#ios)
 
 <a name="clientcerts"></a>
-### Server and client Certifcates (macOS, Android) ###
+### Server and client Certificates (macOS, Android, iOS) ###
```
Comment (alimakki) [inline playbooks/roles/openconnect/templates/instructions.md.j2]: I changed this to be specific to macOS and Android as the p12s are now embedded in the [mobileconfig](https://github.com/StreisandEffect/streisand/blob/d4e3bbc60c40e9e523f562ae91506f65582e5614/playbooks/roles/openconnect/templates/instructions.md.j2#L16).

File: playbooks/roles/test-client/tasks/main.yml [hunk 2]
```diff
@@ -18,124 +18,65 @@
...
-    validate_certs: no
-    mode: 0600
-
-- name: "Remove existing forward SSH config from client if required"
-  file:
-    path: "{{ forward_ssh_config }}"
-    state: absent
-
-- name: "Install an SSH config file for a SSH SOCKS connection with the forward user"
+- name: "Install gateway test script"
```
Comment (alimakki) [inline playbooks/roles/test-client/tasks/main.yml]: Install gateway test script -> Install the gateway test script? Reads a little better IMO. Just a nit.

File: playbooks/roles/test-client/tasks/main.yml [hunk 3]
```diff
@@ -18,124 +18,65 @@
...
-  args:
-    chdir: "/root"
-    creates: "{{ shadowsocks_client }}"
+    mode: 0755
 
-- name: "Set executable mode on Shadowsocks client"
-  file:
-    path: "{{ shadowsocks_client }}"
-    mode: 0700
+- name: "Wait for Streisand gateway"
```
Comment (alimakki) [inline playbooks/roles/test-client/tasks/main.yml]: Wait for Streisand gateway -> Wait for the Streisand gateway

File: playbooks/roles/test-client/tasks/main.yml [hunk 4]
```diff
@@ -18,124 +18,65 @@
...
 
-- name: "Remove existing Shadowsocks QR code from client if required"
-  file:
-    path: "{{ shadowsocks_qr_file }}"
-    state: absent
+- name: "Run the gateway test script"
+  command: "streisand-gateway-test"
 
-- name: "Download the Shadowsocks QR code to the client machine"
+- name: "Download client profile list"
```
Comment (alimakki) [inline playbooks/roles/test-client/tasks/main.yml]: Download client profile list -> Download the client profile list

File: playbooks/roles/test-client/tasks/openconnect-profiletest.yml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      # Use `xmllint` to evaluate an XPATH against the client mobileconfig XML.
+      # The XPATH provided will extract the PKCS12 password for the given profile.
+      shell: >
+        xmllint --xpath '/plist[@version="1.0"]/dict/array/dict[2]//string[1]/text()' \
+                {{ openconnect_profile_dir }}/{{ profile_name }}.mobileconfig
+      register: openconnect_pkcs12_password_contents
+      changed_when: False
+
+    - name: "Start OpenConnect with PKCS12 auth and the {{ profile_name }} profile"
+      # openconnect expects the PKCS12 password to be provided via stdin so we
```
Comment (alimakki) [inline playbooks/roles/test-client/tasks/openconnect-profiletest.yml]: openconnect -> OpenConnect

**Comment by alimakki** [comment]
Running the branch to have a go at all the tests. Will report back shrotly :)

**Comment by alimakki** [comment]
Getting the following errors when testing OpenConnect:

```
TASK [test-client : Start OpenConnect with PKCS12 auth and the planet.author profile] ***
fatal: [streisand-client]: FAILED! => {"changed": true, "cmd": "echo -n 'Configures VPN settingsCertificate{DeviceUid}' | openconnect --background --pid-file /tmp/openconnect.pid --cafile /root/openconnect.ca.crt --certificate /root/openconnect-profiles/planet.author.p12 10.0.0.10:4443", "delta": "0:00:00.019506", "end": "2017-12-07 23:42:03.330118", "msg": "non-zero return code", "rc": 1, "start": "2017-12-07 23:42:03.310612", "stderr": "Enter PKCS#12 pass phrase:\nFailed to decrypt PKCS#12 certificate file\nEnter PKCS#12 pass phrase:\nfgets (stdin): Inappropriate ioctl for device", "stderr_lines": ["Enter PKCS#12 pass phrase:", "Failed to decrypt PKCS#12 certificate file", "Enter PKCS#12 pass phrase:", "fgets (stdin): Inappropriate ioctl for device"], "stdout": "POST https://10.0.0.10:4443/\nAttempting to connect to server 10.0.0.10:4443", "stdout_lines": ["POST https://10.0.0.10:4443/", "Attempting to connect to server 10.0.0.10:4443"]}
```
Is this expected in the current state?

**Comment by cpu** [comment]
> Getting the following errors when testing OpenConnect:

@alimakki Ahh, I think my crummy Xpath was broken by some changes that happened on your mobileconfig branch:

> "echo -n 'Configures VPN settingsCertificate{DeviceUid}'

Looks like it pulled out the wrong thing, not the PKCS12 password. I'll push a fix with a more robust Xpath in the next couple of days. Thanks for flagging that!

File: playbooks/roles/openconnect/templates/instructions.md.j2 (hunk 1 above)
Comment (cpu) [reply playbooks/roles/openconnect/templates/instructions.md.j2]: Fixed in b492c1a

File: playbooks/roles/test-client/tasks/main.yml (hunk 2 above)
Comment (cpu) [reply playbooks/roles/test-client/tasks/main.yml]: Fixed this across the file in 8409df0 - Thanks!

File: playbooks/roles/test-client/tasks/openconnect-profiletest.yml (hunk 5 above)
Comment (cpu) [reply playbooks/roles/test-client/tasks/openconnect-profiletest.yml]: Fixed in 386edd7 - Thanks!

**Comment by cpu** [comment]
> Looks like it pulled out the wrong thing, not the PKCS12 password. I'll push a fix with a more robust Xpath in the next couple of days. Thanks for flagging that!

I wrote a better XPATH in https://github.com/StreisandEffect/streisand/pull/1047/commits/7b766d1eceb00c0653992c80d8837cb87a6a301a - it should withstand minor refactors of the mobileconfig file and I left some comments to help guide the next person that may have to update it.

**Comment by cpu** [comment]
> Getting the following errors when testing OpenConnect:
> TASK [test-client : Start OpenConnect with PKCS12 auth and the planet.author profile] ***
fatal: [streisand-client]: FAILED! => {"changed": true, "cmd": "echo -n 'Configures VPN settingsCertificate{DeviceUid}' | openconnect --background --pid-file /tmp/openconnect.pid --cafile /root/openconnect.ca.crt --certificate /root/openconnect-profiles/planet.author.p12 10.0.0.10:4443", "delta": "0:00:00.019506", "end": "2017-12-07 23:42:03.330118", "msg": "non-zero return code", "rc": 1, "start": "2017-12-07 23:42:03.310612", "stderr": "Enter PKCS#12 pass phrase:\nFailed to decrypt PKCS#12 certificate file\nEnter PKCS#12 pass phrase:\nfgets (stdin): Inappropriate ioctl for device", "stderr_lines": ["Enter PKCS#12 pass phrase:", "Failed to decrypt PKCS#12 certificate file", "Enter PKCS#12 pass phrase:", "fgets (stdin): Inappropriate ioctl for device"], "stdout": "POST https://10.0.0.10:4443/\nAttempting to connect to server 10.0.0.10:4443", "stdout_lines": ["POST https://10.0.0.10:4443/", "Attempting to connect to server 10.0.0.10:4443"]}
> Is this expected in the current state?

@alimakki Should be all fixed now! Can you :mag: again when you have a chance? I was able to run the remote tests against a freshly provisioned instance (with `streisand_client_test: yes`) without error as of 7b766d1

**Comment by cpu** [comment]
I finished an integration test for stunnel. Since this branch isn't approved & merged yet I threw it on the end.

**Comment by cpu** [comment]
The BIP name generation changes broke this branch. Closing for now while I fix the new bugs.

**Comment by cpu** [comment]
This branch is working again now that https://github.com/StreisandEffect/streisand/pull/1108 is merged.

**Comment by cpu** [comment]
@alimakki This should be ready for review again. Thanks!

**Pull request merged**